core/src/apps/ethereum/sign_tx.py

```diff
@@ -101,19 +128,93 @@ async def sign_tx(ctx, msg, keychain):
+    # Handle ERC20s
+    token, address_bytes, recipient, value = await handle_erc20(ctx, msg)
+
+    data_total = msg.data_length
+
+    await require_confirm_tx(ctx, recipient, value, msg.chain_id, token, msg.tx_type)
+    if token is None and msg.data_length > 0:
+        await require_confirm_data(ctx, msg.data_initial_chunk, data_total)
+
+    await require_confirm_eip1559_fee(
+        ctx,
+        int.from_bytes(msg.max_priority_fee, "big"),
+        int.from_bytes(msg.max_gas_fee, "big"),
+        int.from_bytes(msg.gas_limit, "big"),
+        msg.chain_id,
+        msg.tx_type,
+    )
+
+    data = bytearray()
+    data += msg.data_initial_chunk
+    data_left = data_total - len(msg.data_initial_chunk)
+
+    total_length = get_total_length(msg, data_total)
+
+    sha = HashWriter(sha3_256(keccak=True))
+
+    if msg.tx_type is not None:
+        sha.append(msg.tx_type)
+
+    sha.extend(rlp.encode_length(total_length, True))  # total length
+
+    for field in (
+        msg.chain_id,
+        msg.nonce,
+        msg.max_priority_fee,
+        msg.max_gas_fee,
+        msg.gas_limit,
+        address_bytes,
+        msg.value,
+    ):
+        sha.extend(rlp.encode(field))
+
+    if data_left == 0:
+        sha.extend(rlp.encode(data))
+    else:
+        sha.extend(rlp.encode_length(data_total, False))
+        sha.extend(rlp.encode(data, False))
+
+    while data_left > 0:
+        resp = await send_request_chunk(ctx, data_left)
+        data_left -= len(resp.data_chunk)
+        sha.extend(resp.data_chunk)
+
+    access_list = format_access_list(msg.access_list)
+
+    sha.extend(rlp.encode(access_list))
+
+    digest = sha.get_digest()
+    result = sign_digest(msg, keychain, digest)
+
+    return result
+
+
 def get_total_length(msg: EthereumSignTx, data_total: int) -> int:
     length = 0
-    if msg.tx_type is not None:
+    if msg.tx_type is not None and not msg.eip2718:
         length += rlp.field_length(1, [msg.tx_type])
 
     length += rlp.field_length(len(msg.nonce), msg.nonce[:1])
-    length += rlp.field_length(len(msg.gas_price), msg.gas_price)
+    if not msg.eip2718 and msg.gas_price is not None:
+        length += rlp.field_length(len(msg.gas_price), msg.gas_price)
+    if msg.eip2718 and msg.max_gas_fee is not None:
+        length += rlp.field_length(len(msg.max_gas_fee), msg.max_gas_fee)
+    if msg.eip2718 and msg.max_priority_fee is not None:
+        length += rlp.field_length(len(msg.max_priority_fee), msg.max_priority_fee)
+    if msg.eip2718:
+        access_list = format_access_list(msg.access_list)
+        length += len(rlp.encode(access_list))
     length += rlp.field_length(len(msg.gas_limit), msg.gas_limit)
     to = address.bytes_from_address(msg.to)
     length += rlp.field_length(len(to), to)
     length += rlp.field_length(len(msg.value), msg.value)
 
-    if msg.chain_id:  # forks replay protection
+    if msg.eip2718:
+        length += rlp.field_length(1, [msg.chain_id])
```

Comment:
don't you need to figure out the length in the same way as below?

Reply:
Yeah, good catch. I just had issues with extra length you guys were adding for a reason not entirely clear to me 😄